base/array.jl

```diff
@@ -2336,20 +2336,48 @@ Int64[]
 function findall(A)
     collect(first(p) for p in pairs(A) if last(p))
 end
+
 # Allocating result upfront is faster (possible only when collection can be iterated twice)
-function findall(A::AbstractArray{Bool})
-    n = count(A)
+function _findall(f::Function, A::AbstractArray{Bool})
+    # Compute f for true and false only once
+    ft, ff = f(true), f(false)
+    (ft | ff) || return Vector{eltype(keys(A))}()
+    (ft & ff) && return vec(Array(keys(A)))
+    n = let
+        c = count(A)
+        ft ? c : length(A) - c
+    end
     I = Vector{eltype(keys(A))}(undef, n)
+    _findall(ff, I, A)
+end
+
+function _findall(invert::Bool, I::Vector, A::AbstractArray{Bool})
     cnt = 1
-    for (i,a) in pairs(A)
-        if a
-            I[cnt] = i
-            cnt += 1
-        end
+    len = length(I)
+    for (k, v) in pairs(A)
+        cnt > len && break
+        I[cnt] = k
+        cnt += v ⊻ invert
     end
     I
 end
 
+function _findall(invert::Bool, I::Vector, A::AbstractVector{Bool})
+    i = firstindex(A)
+    cnt = 1
+    len = length(I)
+    @inbounds while cnt ≤ len
+        I[cnt] = i
+        cnt += A[i] ⊻ invert
+        i = nextind(A, i)
+    end
+    I
+end
```

Comment:
I assume you have checked that this method for vectors is faster than the generic one for arrays when compiled for vectors? :-)

Reply:
Yep, this is the one the benchmark in the OP times.

Reply:
Yes, but I mean, is it faster than the method for `AbstractArray` that you also introduce in this PR?

Reply:
Ah, I see! Yes, it is faster, but only by about 30%

Reply:
Yeah, maybe we could just drop it. Isn't the performance difference between the two methods due to `@inbounds`?

Though regarding `@inbounds` I think it's considered as OK to assume that `nextind` returns a valid index. I don't understand why @vtjnash said that `@inbounds` isn't OK for the `AbstractArray` method.

Reply:
Redid the benchmark again more carefully - there is actually a significant difference between the vector and array implementations. So keeping it for now.